sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -152,3 +153,56 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
       if (j.joinType == newJoinType) f else Filter(condition, j.copy(joinType = newJoinType))
   }
 }
+
+/**
+ * Correctly handle PythonUDF which need access both side of join side by changing the new join
+ * type to Cross.
+ */
+object HandlePythonUDFInJoinCondition extends Rule[LogicalPlan] with PredicateHelper {
+  def hasPythonUDF(expression: Expression): Boolean = {
+    expression.collectFirst { case udf: PythonUDF => udf }.isDefined
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case j @ Join(_, _, joinType, condition)
+      if condition.map(splitConjunctivePredicates).getOrElse(Nil).exists(hasPythonUDF) =>
+      if (!joinType.isInstanceOf[InnerLike] && joinType != LeftSemi) {
+        // The current strategy only support InnerLike and LeftSemi join because for other type,
+        // it breaks SQL semantic if we run the join condition as a filter after join. If we pass
+        // the plan here, it'll still get a an invalid PythonUDF RuntimeException with message
+        // `requires attributes from more than one child`, we throw firstly here for better
+        // readable information.
+        throw new AnalysisException("Using PythonUDF in join condition of join type" +
+          s" $joinType is not supported.")
+      }
+      if (SQLConf.get.crossJoinEnabled) {
+        // if condition expression contains python udf, it will be moved out from
+        // the new join conditions, and the join type will be changed to CrossJoin.
+        logWarning(s"The join condition:$condition of the join plan contains " +
+          "PythonUDF, it will be moved out and the join plan will be " +
+          s"turned to cross join. This plan shows below:\n $j")
+        val (udf, rest) =
+          condition.map(splitConjunctivePredicates).get.partition(hasPythonUDF)
+        val newCondition = if (rest.isEmpty) {
+          Option.empty
+        } else {
+          Some(rest.reduceLeft(And))
+        }
+        val newJoin = j.copy(joinType = Cross, condition = newCondition)
+        joinType match {
+          case _: InnerLike =>
+            Filter(udf.reduceLeft(And), newJoin)
+          case LeftSemi =>
+            Project(
```

Comment:
we don't need to add the project if we don't change the join type, right?

Reply:
Yes, that's right.